### ISSUE TITLE
`booster.predict(..., pred_leaf = True)` way slower than other flags.

### ISSUE THREAD

**Issue by dacian-dataheroes** [dmlc/xgboost#9725]
Why predicting the leaf is **way** slower than actual predicting? Doesn't xgboost have to get the leaf anyway for the predicted value?

```python
X, y = make_regression(n_samples = 100000, n_features = 10)
dmatrix = xgb.DMatrix(X, label = y)

model = xgb.XGBRegressor(n_estimators = 20, max_depth = 6)
model.fit(X, y)
booster = model.get_booster()

%%timeit 
booster.predict(dmatrix, pred_leaf = True)
 # 22.9 ms ± 626 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit 
booster.predict(dmatrix)
# 385 µs ± 26 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

(Isolating a single booster produces similar results)

**Comment by trivialfis** [comment]
The result for normal prediction is cached, given the same DMatrix, XGBoost can retrieve the result from the previous prediction.

**Comment by dacian-dataheroes** [comment]
Firstly, how does xgboost know it's the same matrix in the above example? I fit it with `X, y`, and predict on my `dmatrix` that's created outsitde. I suppose if xgboost creates a `dmatrix` from `X, y` it will be a different python object, and I doubt xgboost verifies each entry in 2 databases.

Secondly, If I recreate a dataset in before predicting I will still see the difference:

```python
```python
X, y = make_regression(n_samples = 100000, n_features = 10)
dmatrix = xgb.DMatrix(X, label = y)

model = xgb.XGBRegressor(n_estimators = 20, max_depth = 6)
model.fit(X, y)
booster = model.get_booster()

X, y = make_regression(n_samples = 100000, n_features = 10)
dmatrix = xgb.DMatrix(X, label = y)

%%timeit 
booster.predict(dmatrix, pred_leaf = True)
 # 20 ms ± 518 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%%timeit 
booster.predict(dmatrix)
# 424 µs ± 14.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by dacian-dataheroes** [comment]
Maybe I'm still misunderstanding, but `X` from `make_regression` is of dtype f64, `dmatrix.get_data()` returns a sparse format of f32
```python
X.dtype
# dtype('float64')

dmatrix.get_data()
# <100000x10 sparse matrix of type '<class 'numpy.float32'>'
# 	with 1000000 stored elements in Compressed Sparse Row format>
```
This either means that when I call `get_data()` I convert the saved f64 data into f32 for some reason, or it's saved in f32. If it's the case of being saved in f32, then a new copy is made inside the booster, right? And when I create the dmatrix outside then another one is made there. This means I have 2 different sparsef32 matrices.

I still get different results with a new dataset.

(I haven't gone into the C++ code, only played with python / sklearn)

**Comment by trivialfis** [comment]
It's being copied as f32. XGB handles only f32, get data makes another copy in scipy csr. If you want a prediction without cache, use inplace predict from booster instead.

**Comment by dacian-dataheroes** [comment]
In the end, I want to get the speed of `.predict()` when I do `.predict(pred_leaf = True)`, if posible. 

>  If you want a prediction without cache, use inplace predict from booster instead.

I don't understand how this is relevant. If it cached the predictions and just returned them, then I would get the same predictions no matter the data in the dmatrix. However I'm getting different predictions for different dmatrices, and they are fast.

**Comment by trivialfis** [comment]
``` python
m = DMatrix(X)
booster.predict(m) # first prediction, goes through all trees, no cache
booster.predict(m) # second call, uses cached value
m = DMatrix(X)
booster.predict(m) # DMatrix is re-created, goes through all trees, no cache
```

**Comment by dacian-dataheroes** [comment]
> lastly, %%timeit runs the same line multiple times to get the variance, except for the first call all the following calls are cached. 

I see, good point, my bad here. Using `time` would help? I also isolated the last booster in `booster19` to do predictions with. 

```python
X, y = make_regression(n_samples = 100000, n_features = 10)
model = xgb.XGBRegressor(n_estimators = 20, max_depth = 6)
model.fit(X, y)
booster = model.get_booster()
booster19 = booster[19]

t_list = []
for t in range(100):
    X, y = make_regression(n_samples = 100000, n_features = 10)
    dmatrix = xgb.DMatrix(X, label = y)
    t = time()
    booster19.predict(dmatrix, pred_leaf = True)
    t_list.append(time() - t)
 
print(np.mean(t_list) * 1000)
# 4.378945827484131

t_list = []
for t in range(100):
    X, y = make_regression(n_samples = 100000, n_features = 10)
    dmatrix = xgb.DMatrix(X, label = y)
    t = time()    
    booster19.predict(dmatrix, pred_leaf = True)
    t_list.append(time() - t)

print(np.mean(t_list) * 1000)
#3.6478185653686523


t_list = []
for t in range(100):
    X, y = make_regression(n_samples = 100000, n_features = 10)
    dmatrix = xgb.DMatrix(X, label = y)
    t = time()
    booster19.predict(dmatrix)
    t_list.append(time() - t)
    
print(np.mean(t_list) * 1000)
# 1.9671010971069338
```

I still get the difference, although smaller. However, intuitively don't you have to go to the leaf first to get the prediction value? Which means they should have similar times, however `pred_leaf = True` is consistently slower.

**Comment by dacian-dataheroes** [comment]
> lastly, %%timeit runs the same line multiple times to get the variance, except for the first call all the following calls are cached. 

I see, good point, my bad here. Using `time` would help? I also isolated the last booster in `booster19` to do predictions with. 

```python
X, y = make_regression(n_samples = 100000, n_features = 10)
model = xgb.XGBRegressor(n_estimators = 20, max_depth = 6)
model.fit(X, y)
booster = model.get_booster()
booster19 = booster[19]

t_list = []
for t in range(100):
    X, y = make_regression(n_samples = 100000, n_features = 10)
    dmatrix = xgb.DMatrix(X, label = y)
    t = time()    
    booster19.predict(dmatrix, pred_leaf = True)
    t_list.append(time() - t)

print(np.mean(t_list) * 1000)
#3.6478185653686523


t_list = []
for t in range(100):
    X, y = make_regression(n_samples = 100000, n_features = 10)
    dmatrix = xgb.DMatrix(X, label = y)
    t = time()
    booster19.predict(dmatrix)
    t_list.append(time() - t)
    
print(np.mean(t_list) * 1000)
# 1.9671010971069338
```

I still get the difference, although smaller. However, intuitively don't you have to go to the leaf first to get the prediction value? Which means they should have similar times, however `pred_leaf = True` is consistently slower. 

Putting them in the same for loop (to have the same dataset) yielded similar results

```python
t_list1 = []
t_list2 = []
for t in range(100):
    X, y = make_regression(n_samples = 100000, n_features = 10)
    dmatrix = xgb.DMatrix(X, label = y)
    
    t = time()
    booster19.predict(dmatrix)
    t_list1.append(time() - t)
    
    t = time()    
    booster19.predict(dmatrix, pred_leaf = True)
    t_list2.append(time() - t)

print(np.mean(t_list1) * 1000)
print(np.mean(t_list2) * 1000)
# 1.6228461265563965
# 3.3811211585998535

```
